src/datatypes/list.h

```diff
@@ -0,0 +1,267 @@
+/**
+* @file datatypes/list.h
+*
+* @brief List datatype
+*
+* A generic doubly-linked list
+*
+* SPDX-FileCopyrightText: 2008-2022 HPDCS Group <rootsim@googlegroups.com>
+* SPDX-License-Identifier: GPL-3.0-only
+*/
+#pragma once
+
+#include <assert.h>
+#include <memory.h>
+#include <stddef.h>
+
+/// This structure defines a generic list. Nodes of the list must have a next/prev pointer properly typed.
+struct list {
+	/// the size of the list
+	size_t size;
+	/// pointer to the first element
+	void *head;
+	/// pointer to the last element
+	void *tail;
+};
+
+/// This macro is a slightly-different implementation of the standard offsetof macro
+#define my_offsetof(st, m) ((size_t)( (unsigned char *)&((st)->m ) - (unsigned char *)(st)))
+
+/// Declare a "typed" list. This is a pointer to type, but the variable will instead reference a struct rootsim_list!
+#define list(type) type *
+
+/** This macro allocates a struct rootsim_list object and cast it to the type pointer.
+ *  It can be used to mimic the C++ syntax of templated lists, like:
+ *  \code
+ *   list(int) = new_list(int);
+ *  \endcode
+ */
+#define new_list(type)	\
+	__extension__({ \
+		void *__lmptr; \
+		__lmptr = malloc(sizeof(struct list)); \
+		memset(__lmptr, 0, sizeof(struct list));\
+		__lmptr;\
+	})
+
+// Get the size of the current list.
+#define list_sizeof(list) ((struct list *)list)->size
+
+/**
+ * This macro retrieves a pointer to the payload of the head node of a list.
+ *
+ * @param li a pointer to a list created using the new_list() macro.
+ */
+#define list_head(li) ((__typeof__ (li))(((struct list *)(li))->head))
+
+/**
+ * This macro retrieves a pointer to the payload of the tail node of a list.
+ *
+ * @param li a pointer to a list created using the new_list() macro.
+ */
+#define list_tail(li) ((__typeof__ (li))(((struct list *)(li))->tail))
+
+/**
+ * Given a pointer to a list node's payload, this macro retrieves the next node's payload, if any.
+ *
+ * @param ptr a pointer to a list created using the new_list() macro.
+ */
+#define list_next(ptr) ((ptr)->next)
+
+/**
+ * Given a pointer to a list node's payload, this macro retrieves the prev node's payload, if any.
```

Comment:
Same as for line 51, for clarity
```suggestion
 * Given a pointer to a list node, this macro retrieves a pointer to the prev node, if any.
```